File: frontend/ios/VehicleReport.swift
import Foundation

// MARK: - Main Report

/// Complete Vehicle Report — the main API response from POST /api/v1/vehicles/query.
struct VehicleReport: Codable {
    let plate: String
    let cached: Bool
    let generatedAt: String
    let basic: BasicInfo
    let ownership: OwnershipInfo
    let inspection: InspectionInfo
    let recalls: RecallInfo
    let tax: TaxInfo
    let risk: RiskInfo
    let priceEstimate: PriceEstimate
    let conditionInference: ConditionInference
    let suitability: Suitability
    let aiSummary: String

    enum CodingKeys: String, CodingKey {
        case plate, cached, basic, ownership, inspection, recalls, tax, risk, suitability
        case generatedAt = "generated_at"
        case priceEstimate = "price_estimate"
        case conditionInference = "condition_inference"
        case aiSummary = "ai_summary"
    }
}

// MARK: - Basic Info

struct BasicInfo: Codable {
    let plate: String
    let vin: String?
    let make: String
    let model: String
    let variant: String?
    let year: Int
    let bodyType: String
    let fuelType: String
    let engineCode: String?
    let engineDisplacementL: Double?
    let enginePowerHp: Int?
    let engineTorqueNm: Int?
    let gearboxType: String?
    let gearboxGears: Int?
    let drivetrainType: String?
    let is4x4: Bool
    let color: String?
    let seats: Int?
    let curbWeightKg: Int?
    let firstRegistration: String?
    let countryOrigin: String?
    let mileageLatest: Int?
    let ownerCount: Int?

    enum CodingKeys: String, CodingKey {
        case plate, vin, make, model, variant, year, color, seats
        case bodyType = "body_type"
        case fuelType = "fuel_type"
        case engineCode = "engine_code"
        case engineDisplacementL = "engine_displacement_l"
        case enginePowerHp = "engine_power_hp"
        case engineTorqueNm = "engine_torque_nm"
        case gearboxType = "gearbox_type"
        case gearboxGears = "gearbox_gears"
        case drivetrainType = "drivetrain_type"
        case is4x4 = "is_4x4"
        case curbWeightKg = "curb_weight_kg"
        case firstRegistration = "first_registration"
        case countryOrigin = "country_origin"
        case mileageLatest = "mileage_latest"
        case ownerCount = "owner_count"
    }
}

// MARK: - Ownership

struct OwnershipInfo: Codable {
    let ownerCount: Int
    let history: [OwnershipEvent]

    enum CodingKeys: String, CodingKey {
        case ownerCount = "owner_count"
        case history
    }
}

struct OwnershipEvent: Codable {
    let date: String
    let event: String
}

// MARK: - Inspection

struct InspectionInfo: Codable {
    let totalInspections: Int
    let lastInspection: String?
    let lastResult: String?
    let history: [Inspection]

    enum CodingKeys: String, CodingKey {
        case totalInspections = "total_inspections"
        case lastInspection = "last_inspection"
        case lastResult = "last_result"
        case history
    }
}

struct Inspection: Codable {
    let date: String
    let mileageKm: Int
    let result: String
    let notes: String?

    enum CodingKeys: String, CodingKey {
        case date, result, notes
        case mileageKm = "mileage_km"
    }
}

// MARK: - Recalls

struct RecallInfo: Codable {
    let totalRecalls: Int
    let openRecalls: Int
    let recalls: [Recall]

    enum CodingKeys: String, CodingKey {
        case totalRecalls = "total_recalls"
        case openRecalls = "open_recalls"
        case recalls
    }
}

struct Recall: Codable {
    let vin: String
    let source: String
    let description: String
    let severity: String
    let fixAvailable: Bool

    enum CodingKeys: String, CodingKey {
        case vin, source, description, severity
        case fixAvailable = "fix_available"
    }
}

// MARK: - Tax

struct TaxInfo: Codable {
    let annualTaxSek: Int
    let co2EmissionGKm: Int
    let euroStandard: String

    enum CodingKeys: String, CodingKey {
        case annualTaxSek = "annual_tax_sek"
        case co2EmissionGKm = "co2_emission_g_km"
        case euroStandard = "euro_standard"
    }
}

// MARK: - Risk

struct RiskInfo: Codable {
    let score: Int
    let level: String
    let factors: [String]
    let description: String
}

// MARK: - Price Estimate

struct PriceEstimate: Codable {
    let minSek: Int
    let maxSek: Int
    let medianSek: Int
    let currency: String
    let source: String

    enum CodingKeys: String, CodingKey {
        case currency, source
        case minSek = "min_sek"
        case maxSek = "max_sek"
        case medianSek = "median_sek"
    }
}

// MARK: - Condition Inference

struct ConditionInference: Codable {
    let rating: String
    let confidence: Double
    let indicators: [String]
    let description: String
}

// MARK: - Suitability

struct Suitability: Codable {
    let score: Int
    let recommendation: String
    let pros: [String]
    let cons: [String]
}

// MARK: - User

struct CreditBalance: Codable {
    let userId: Int
    let plan: String
    let credits: Int

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case plan, credits
    }
}

struct QueryHistoryItem: Codable {
    let plate: String
    let queriedAt: String
    let cached: Bool

    enum CodingKeys: String, CodingKey {
        case plate, cached
        case queriedAt = "queried_at"
    }
}

File: frontend/android/VehicleReport.kt
package com.cardeal.models

import com.google.gson.annotations.SerializedName

/**
 * Complete Vehicle Report — the main API response from POST /api/v1/vehicles/query.
 */
data class VehicleReport(
    val plate: String,
    val cached: Boolean,
    @SerializedName("generated_at") val generatedAt: String,
    val basic: BasicInfo,
    val ownership: OwnershipInfo,
    val inspection: InspectionInfo,
    val recalls: RecallInfo,
    val tax: TaxInfo,
    val risk: RiskInfo,
    @SerializedName("price_estimate") val priceEstimate: PriceEstimate,
    @SerializedName("condition_inference") val conditionInference: ConditionInference,
    val suitability: Suitability,
    @SerializedName("ai_summary") val aiSummary: String
)

data class BasicInfo(
    val plate: String,
    val vin: String? = null,
    val make: String,
    val model: String,
    val variant: String? = null,
    val year: Int,
    @SerializedName("body_type") val bodyType: String,
    @SerializedName("fuel_type") val fuelType: String,
    @SerializedName("engine_code") val engineCode: String? = null,
    @SerializedName("engine_displacement_l") val engineDisplacementL: Double? = null,
    @SerializedName("engine_power_hp") val enginePowerHp: Int? = null,
    @SerializedName("engine_torque_nm") val engineTorqueNm: Int? = null,
    @SerializedName("gearbox_type") val gearboxType: String? = null,
    @SerializedName("gearbox_gears") val gearboxGears: Int? = null,
    @SerializedName("drivetrain_type") val drivetrainType: String? = null,
    @SerializedName("is_4x4") val is4x4: Boolean = false,
    val color: String? = null,
    val seats: Int? = null,
    @SerializedName("curb_weight_kg") val curbWeightKg: Int? = null,
    @SerializedName("first_registration") val firstRegistration: String? = null,
    @SerializedName("country_origin") val countryOrigin: String? = null,
    @SerializedName("mileage_latest") val mileageLatest: Int? = null,
    @SerializedName("owner_count") val ownerCount: Int? = null
)

data class OwnershipInfo(
    @SerializedName("owner_count") val ownerCount: Int,
    val history: List<OwnershipEvent> = emptyList()
)

data class OwnershipEvent(
    val date: String,
    val event: String
)

data class InspectionInfo(
    @SerializedName("total_inspections") val totalInspections: Int,
    @SerializedName("last_inspection") val lastInspection: String? = null,
    @SerializedName("last_result") val lastResult: String? = null,
    val history: List<Inspection> = emptyList()
)

data class Inspection(
    val date: String,
    @SerializedName("mileage_km") val mileageKm: Int,
    val result: String,
    val notes: String? = null
)

data class RecallInfo(
    @SerializedName("total_recalls") val totalRecalls: Int,
    @SerializedName("open_recalls") val openRecalls: Int,
    val recalls: List<Recall> = emptyList()
)

data class Recall(
    val vin: String,
    val source: String,
    val description: String,
    val severity: String,
    @SerializedName("fix_available") val fixAvailable: Boolean
)

data class TaxInfo(
    @SerializedName("annual_tax_sek") val annualTaxSek: Int,
    @SerializedName("co2_emission_g_km") val co2EmissionGKm: Int,
    @SerializedName("euro_standard") val euroStandard: String
)

data class RiskInfo(
    val score: Int,
    val level: String,
    val factors: List<String> = emptyList(),
    val description: String
)

data class PriceEstimate(
    @SerializedName("min_sek") val minSek: Int,
    @SerializedName("max_sek") val maxSek: Int,
    @SerializedName("median_sek") val medianSek: Int,
    val currency: String,
    val source: String
)

data class ConditionInference(
    val rating: String,
    val confidence: Double,
    val indicators: List<String> = emptyList(),
    val description: String
)

data class Suitability(
    val score: Int,
    val recommendation: String,
    val pros: List<String> = emptyList(),
    val cons: List<String> = emptyList()
)

/**
 * Credit balance response from GET /api/v1/user/credits.
 */
data class CreditBalance(
    @SerializedName("user_id") val userId: Int,
    val plan: String,
    val credits: Int
)

/**
 * Query history item from GET /api/v1/user/history.
 */
data class QueryHistoryItem(
    val plate: String,
    @SerializedName("queried_at") val queriedAt: String,
    val cached: Boolean
)
